nbgrader/docs/source/exchange/exchange_api.rst

```diff
@@ -0,0 +1,389 @@
+
+The exchange is called thus::
+
+    exchange = ExchangeFactory(config=config)
+    release = exchange.ReleaseAssignment(
+        coursedir=self.coursedir,
+        authenticator=self.authenticator,
+        parent=self)
+    try:
+        release.start()
+    except ExchangeError:
+        self.fail(``nbgrader release_assignment failed``)
+
+returns.... *nothing*
+
+``ExchangeReleaseFeedback``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This should copy all the feedback for the current assignment to the exchange.
+
+Feedback is generated by the Instructor. From ``GenerateFeedbackApp``::
+
+    Create HTML feedback for students after all the grading is finished.
+    This takes a single parameter, which is the assignment ID, and then (by
+    default) looks at the following directory structure:
+
+        autograded/*/{assignment_id}/*.ipynb
+
+    from which it generates feedback the the corresponding directories
+    according to:
+
+        feedback/{student_id}/{assignment_id}/{notebook_id}.html
+
+The exchange is called thus::
+
+    exchange = ExchangeFactory(config=config)
+    release_feedback = exchange.ReleaseFeedback(
+        coursedir=self.coursedir,
+        authenticator=self.authenticator,
+        parent=self)
+    try:
+        release_feedback.start()
+    except ExchangeError:
+        self.fail("nbgrader release_feedback failed")
+
+returns..... nothing
+
+``ExchangeSubmit``
+~~~~~~~~~~~~~~~~~~
+
+This should copy the assignment from the user's work space, and make it available for instructors to *collect*.
+
+The exchange is called thus::
+
+    with self.get_assignment_dir_config() as config:
+        try:
+            config = self.load_config()
+            config.CourseDirectory.course_id = course_id
+            config.CourseDirectory.assignment_id = assignment_id
+            coursedir = CourseDirectory(config=config)
+            authenticator = Authenticator(config=config)
+            exchange = ExchangeFactory(config=config)
+            submit = exchange.Submit(
+                coursedir=coursedir,
+                authenticator=authenticator,
+                config=config)
+            submit.start()
+        .....
+
+The *source* for files to be submitted needs to match that in ``ExchangeFetchAssignment``.
+
+returns.... *nothing*
+
+When writing your own Exchange
+++++++++++++++++++++++++++++++
+
+- You to need to consider stopping students from seeing each others submissions
+- nbgrader functionality requires a file called ``timestamp.txt`` to be in the submission, containing the timestamp of that submission. The creation of this file is the responsibility of this class.
+- Whilst nothing is done *as yet*, the default exchange checks the names of submitted notebooks, and logs differences.
+- Submissions need to record ``student_id``, as well as ``course_id`` & ``assignment_id``
+- The default exchange copies files to both an ``inbound`` and ``cache`` store. This may be significant considering ``ExchangeList``
```

Comment:
Did the last sentence get cut off? If not, explain why it is significant.

Reply:
I have no idea what happened to that sentence. I've removed it for now as I'm not sure what is supposed to be significant here.

Reply:
The `cached` flag is used in base `nbgrader` to get the list of files in the submission cache (see https://github.com/jupyter/nbgrader/blob/dd0d66a41685c911971b32cbc3243659b2813a39/nbgrader/exchange/list.py#L20)